### ISSUE TITLE
Add metrics for S3 sstables backend

### ISSUE THREAD

**Issue by xemul** [scylladb/scylla-monitoring#2103]
The metrics in question partially mimic those for IO classes:
* `scylla_s3_total_read_requests` is that same as `scylla_io_queue_total_read_ops`
* `scylla_s3_total_write_requests` is that same as `scylla_io_queue_total_write_ops`
* `scylla_s3_total_read_bytes` is that same as `scylla_io_queue_total_read_bytes`
* `scylla_s3_total_write_bytes` is that same as `scylla_io_queue_total_write_bytes`

Partially extends them
* `scylla_s3_total_read_latency_sec` is that same as `scylla_io_queue_total_delay_sec` but only for reads
* `scylla_s3_total_write_latency_sec` is that same as `scylla_io_queue_total_delay_sec` but only for writes

(the above latencies should be divided by corresponding *_requests to show per-request latency (like in #1714))

And partially introduces its own
* `scylla_s3_nr_connections` gauge shows the total number of established connections
* `scylla_s3_nr_active_connections` gauge shows the total number of established connections that are serving http request. Respectively `nr_connections - nr_active_connections` value would show the number of connections in pool
* `scylla_s3_total_new_connections` counter shows the number of newly established connections. The faster it grows, the worse, because it means that old connections are dropped for whatever reason and client opens new ones, which is costly

All metrics are per-{scheduling-class , target-endpoint} pair. In most of the cases there will be just one "endpoint" label value.

**Comment by amnonh** [comment]
WhatsApp scylladb versions this is applicable to? OS and enterprise

**Comment by xemul** [comment]
It's going to be in 5.4 (under --experimental option)

**Comment by amnonh** [comment]
And what enterprise version?

**Comment by xemul** [comment]
> And what enterprise version?

I think none, even 5.4 is not released yet

**Comment by amnonh** [comment]
@xemul it must be part of a future enterprise release

**Comment by xemul** [comment]
Yes, then it's likely going to be the 2024.1 version?

**Comment by mykaul** [comment]
> Yes, then it's likely going to be the 2024.1 version?

Yes.

**Comment by amnonh** [comment]
@xemul how can I test this?

**Comment by xemul** [comment]
@amnonh , here's the doc how to set Scylla for that:
https://github.com/scylladb/scylladb/blob/master/docs/dev/object_storage.md

You'll also need to either create a bucket on AWS S3, or start minio server

**Comment by amnonh** [comment]
@xemul I tried to test it with scylladb running on docker 5.4.0-rc1 with minio running in a container

No metrics were created, so I tried to run ```nodetool flush``` and got:
```
INFO  2023-11-08 11:03:27,301 [shard 0:main] commitlog_replayer - Replaying /var/lib/scylla/commitlog/CommitLog-2-153046609.log, /var/lib/scylla/commitlog/CommitLog-2-153046608.log, /var/lib/scylla/commitlog/CommitLog-2-153041438.log, /var/lib/scylla/commitlog/Recycled-CommitLog-2-152480867.log, /var/lib/scylla/commitlog/CommitLog-2-153051779.log, /var/lib/scylla/commitlog/CommitLog-2-152480869.log, /var/lib/scylla/commitlog/CommitLog-2-152480863.log, /var/lib/scylla/commitlog/Recycled-CommitLog-2-152480868.log, /var/lib/scylla/commitlog/Recycled-CommitLog-2-152480866.log, /var/lib/scylla/commitlog/CommitLog-2-153051780.log, /var/lib/scylla/commitlog/CommitLog-2-152480864.log, /var/lib/scylla/commitlog/CommitLog-2-153041439.log
INFO  2023-11-08 11:03:27,356 [shard 0:main] commitlog_replayer - Log replay complete, 23778 replayed mutations (0 invalid, 0 skipped)
INFO  2023-11-08 11:03:27,356 [shard 0:main] init - replaying commit log - flushing memtables
ERROR 2023-11-08 11:03:27,357 [shard 0:main] table - failed to write sstable /var/lib/scylla/data/testing/prepared-30dfb1007e2511eeab867459b658559d/me-3gaw_0upr_24awx2dwafnblnqrel-big-Data.db: seastar::httpd::unexpected_status_error (Unexpected reply status)
ERROR 2023-11-08 11:03:27,357 [shard 0:main] table - Memtable flush failed due to: seastar::httpd::unexpected_status_error (Unexpected reply status). Aborting, at 0x5f987ae 0x5f98d70 0x5f99048 0x1c10c16 0x138d51a 0x5a9bb0f 0x5a9cde7 0x5a9c159 0x5a3ea37 0x5a3dbec 0x1310e4e 0x13128b0 0x130f3bc /opt/scylladb/libreloc/libc.so.6+0x27b89 /opt/scylladb/libreloc/libc.so.6+0x27c4a 0x130cde4
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::future<void>::handle_exception<replica::dirty_memory_manager::flush_one(replica::memtable_list&, replica::flush_permit&&)::$_0>(replica::dirty_memory_manager::flush_one(replica::memtable_list&, replica::flush_permit&&)::$_0&&)::{lambda(auto:1&&)#1}, seastar::future<void>::then_wrapped_nrvo<seastar::future<void>, seastar::future<void>::handle_exception<replica::dirty_memory_manager::flush_one(replica::memtable_list&, replica::flush_permit&&)::$_0>(replica::dirty_memory_manager::flush_one(replica::memtable_list&, replica::flush_permit&&)::$_0&&)::{lambda(auto:1&&)#1}>(seastar::future<void>::handle_exception<replica::dirty_memory_manager::flush_one(replica::memtable_list&, replica::flush_permit&&)::$_0>(replica::dirty_memory_manager::flush_one(replica::memtable_list&, replica::flush_permit&&)::$_0&&)::{lambda(auto:1&&)#1}&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::future<void>::handle_exception<replica::dirty_memory_manager::flush_one(replica::memtable_list&, replica::flush_permit&&)::$_0>(auto:1&&)::{lambda(auto:1&&)#1}&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::future<void>::finally_body<replica::memtable_list::flush()::$_0::operator()<replica::flush_permit>(replica::flush_permit) const::{lambda()#1}, false>, seastar::future<void>::then_wrapped_nrvo<seastar::future<void>, seastar::future<void>::finally_body<replica::memtable_list::flush()::$_0::operator()<replica::flush_permit>(replica::flush_permit) const::{lambda()#1}, false> >(seastar::future<void>::finally_body<replica::memtable_list::flush()::$_0::operator()<replica::flush_permit>(replica::flush_permit) const::{lambda()#1}, false>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::future<void>::finally_body<replica::memtable_list::flush()::$_0::operator()<replica::flush_permit>(auto:1) const::{lambda()#1}, false>&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::shared_future<>::shared_state::get_future(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::{lambda(seastar::future<void>&&)#1}, seastar::future<void>::then_wrapped_nrvo<void, seastar::shared_future<>::shared_state::get_future(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::{lambda(seastar::future<void>&&)#1}>(seastar::shared_future<>::shared_state::get_future(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::{lambda(seastar::future<void>&&)#1}&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::shared_future<>::shared_state::get_future(std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::{lambda(seastar::future<void>&&)#1}&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
Aborting on shard 0.
Backtrace:
  0x5a8a258
  0x5ac0242
  /opt/scylladb/libreloc/libc.so.6+0x3dbaf
  /opt/scylladb/libreloc/libc.so.6+0x8e883
  /opt/scylladb/libreloc/libc.so.6+0x3dafd
  /opt/scylladb/libreloc/libc.so.6+0x2687e
  0x1c10c49
  0x138d51a
  0x5a9bb0f
  0x5a9cde7
  0x5a9c159
  0x5a3ea37
  0x5a3dbec
  0x1310e4e
  0x13128b0
  0x130f3bc
  /opt/scylladb/libreloc/libc.so.6+0x27b89
  /opt/scylladb/libreloc/libc.so.6+0x27c4a
  0x130cde4
2023-11-08 11:03:30,441 INFO exited: scylla (terminated by SIGABRT (core dumped); not expected)

```

**Comment by xemul** [comment]
> `INFO  2023-11-08 11:03:27,301 [shard 0:main] commitlog_replayer - Replaying ...`

It's a message from early boot, AFAIK, not from flush. Do you have full log?
Other than that -- this typically arrives when minio or object_storage.yaml is misconfigured. You can get more elaborated error code into logs with `--logger-log-level http=debug` option (until scylladb/seastar#1931)

**Comment by amnonh** [comment]
I'm postponding it untill we'll have a stable version with a step-by-step instruction of how to test it, or even better, when it will be covered by QA tests

**Comment by amnonh** [comment]
@xemul I would be happy to bring it back to monitoring 4.6 but there should be a better way to test it

**Comment by amnonh** [comment]
@xemul ping

**Comment by xemul** [comment]
@amnonh , the issue you stepped on [here](https://github.com/scylladb/scylla-monitoring/issues/2103#issuecomment-1801663580) is not something we fixed explicitly. Could you provide more details on what the problem was?

Also

>  ... when it will be covered by QA tests

there's a unit test that starts scylla and populates it with S3-backed data -- `pytest test/object_store/test_basic.py`, would that work for you?

**Comment by amnonh** [comment]
@xemul I just need a way to see those metrics in action, if you can add a step-by-step instructions on how to get those metrics, it's good enough

**Comment by xemul** [comment]
@amnonh , the https://github.com/scylladb/scylladb/blob/master/docs/dev/object_storage.md is the instructions. Me, @tchaikov and object_store test all follow them :) I know that you tried them too and it ended up with unexpected reply, and I'm ready to help debugging it. The `--logger-log-level http=debug` and full logs (as I wrote -- the commitlog replaying message is not from flush) are to start with

**Comment by amnonh** [comment]
@xemul I'm refering to this: ```terminated by SIGABRT (core dumped)``` when I'm trying to write to minio
I can try using AWS bucket, if there is a test QA are already using and I can get metrics samples from there it's fine.

If possible, I prefer not doing testing, but use a working code

**Comment by xemul** [comment]
> @xemul I'm refering to this: `terminated by SIGABRT (core dumped)` when I'm trying to write to minio I can try using AWS bucket, if there is a test QA are already using and I can get metrics samples from there it's fine.

That's still about misconfiguration. In recent version of scylla instead of SIGABRT you'll get a message about inability to flush and scheduled attempt to write later (scylladb/scylladb#13745), but that won't help much I think

> If possible, I prefer not doing testing, but use a working code

It should properly configured and there are tests that prove it works if such. If you can show me a cluster that's not working I'd be happy to check it and fix configuration for you

**Comment by amnonh** [comment]
@xemul is there a QA test we can run?

**Comment by amnonh** [comment]
@xemul I would like it to be part of 4.8 is there a QA test I can use to run and get some metrics examples?

**Comment by xemul** [comment]
@amnonh , nothing had changed since [last time](https://github.com/scylladb/scylla-monitoring/issues/2103#issuecomment-1838274494). There's a unit-test in the tree, but not more than that

**Comment by amnonh** [comment]
@xemul @mykaul, those are 11 new metrics; who will use them, and how?
The dashboards are already too crowded. Do we need all those metrics? Is there an actionable outcome from looking at them? Is it meaningful for the user? If so, how?

**Comment by mykaul** [comment]
> @xemul @mykaul, those are 11 new metrics; who will use them, and how? The dashboards are already too crowded. Do we need all those metrics? Is there an actionable outcome from looking at them? Is it meaningful for the user? If so, how?

I have no doubt they will be immensely useful when S3 storage will be used. If it's per keyspace/table property, perhaps it should be there... ? It could explain slowness, for example.

**Comment by amnonh** [comment]
@mykaul All 11 panels? If it's helpful, how? what should the user look for, and what should they do given the findings?

**Comment by mykaul** [comment]
> @mykaul All 11 panels? If it's helpful, how? what should the user look for, and what should they do given the findings?

In the initial release(s), I don't know what metrics will find useful.

**Comment by amnonh** [comment]
that's my point, how will the user know what to do with it?

**Comment by xemul** [comment]
There are IO-queue metrics on the Advanced dashboard that we use to see how IO happens. The S3 metrics have exactly the same meaning, they show details about IO, but this time not on local disks, but on remote S3 bucket

**Comment by amnonh** [comment]
I have tried to use s3 storage, this time with a bucket and fresh credentials I created. Using Scylla latest running inside a container, I get the following:
```
ERROR 2024-07-03 15:03:15,895 [shard 0:main] table - failed to write sstable /var/lib/scylla/data/ks/monkeyspecies-3ac7c390394d11ef9e39463f9ae5bc86/md-3ghi_15tf_2sy9c2emczln8c65c6-big-Data.db: storage_io_error (S3 error (seastar::tls::verification_error (The certificate is NOT trusted. The certificate issuer is unknown. (Issuer=[C=US,O=Amazon,CN=Amazon RSA 2048 M01], Subject=[CN=*.s3.us-east-2.amazonaws.com]))))
ERROR 2024-07-03 15:03:15,895 [shard 0:strm] table - Memtable flush failed due to: storage_io_error (S3 error (seastar::tls::verification_error (The certificate is NOT trusted. The certificate issuer is unknown. (Issuer=[C=US,O=Amazon,CN=Amazon RSA 2048 M01], Subject=[CN=*.s3.us-east-2.amazonaws.com])))). Aborting, at 0x647476e 0x6474d80 0x6475068 0x1d08201 0x144faea 0x5f6d41f 0x5f6e707 0x5f6da68 0x5efba17 0x5efabdc 0x13dfca8 0x13e16f0 0x13de279 /opt/scylladb/libreloc/libc.so.6+0x27b89 /opt/scylladb/libreloc/libc.so.6+0x27c4a 0x13db664
   --
```

I suggest we'll do the following, run any kind of test that works for you (I just want the metric) copy the prometheus directory and send it to me.

**Comment by amnonh** [comment]
@xemul please see if you can help

**Comment by xemul** [comment]
It looks like you've stepped on https://github.com/scylladb/scylladb/issues/13904

**Comment by amnonh** [comment]
@xemul can you make a simple test (any test is good) work? just tar the Prometheus data and I'll take it from there

**Comment by xemul** [comment]
By "simple test" you likely mean some test with scylla monitoring container running aside?

**Comment by amnonh** [comment]
@xemul I mean any test that would use the S3 metrics, I don't care what the test is doing, I just need the metrics. Just collect the Prometheus directory and send it to me

**Comment by amnonh** [comment]
@xemul I'm going to branch 4.8 this week. I'd rather this issue be part of it, but no metrics, no issue, and it will wait for a future release when I'll have the metrics.

**Comment by xemul** [comment]
I don't quite understand what "Prometheus directory" you refer to. You ask for "any test that would use the S3 metrics", this is what object_store unit test from Scylla directory does from my perspective -- it starts minio server, puts sstables there, so Scylla unavoidably generates S3 metrics if asked for

**Comment by amnonh** [comment]
I gave up on running it myself, so I'm passing the running test to you. If you start the monitoring stack with the ```-d``` command line flag, it stores the Prometheus data in an external directory; once done, tar/gz the result and send it to me.

**Comment by regevran** [comment]
@kreuzerkrieg , it is now yours.

**Comment by tchaikov** [comment]
@kreuzerkrieg hi Ernest, are you still working on this ? @regevran wanted me to take a look at it. I suggest we collaborate to move this forward efficiently.

**Comment by tchaikov** [comment]
Ernest pointed me to https://github.com/scylladb/scylladb/issues/13904. so this issue is pending on it.

**Comment by amnonh** [comment]
hey @regevran I want to include this in the next 4.9 monitoring release. Can you have someone set a test so that I can get the generated metrics?

**Comment by regevran** [comment]
> Can you have someone set a test 

I'll try.
But these metrics have a lower priority at the moment, so I doubt we'll get to it.
I'll do my best.

**Comment by amnonh** [comment]
@regevran if the code works, it shouldn't be very difficult, I'm sure you have working testing for it

**Comment by regevran** [comment]
[Waiting](https://github.com/scylladb/scylladb/issues/13904#issuecomment-2581700312).

**Comment by amnonh** [comment]
So now that I have the metrics, I see that we don't display their equivalent, so I think we should show something but probably not all the metrics

**Comment by regevran** [comment]
> I see that we don't display their equivalent

Can you please elaborate?

S3 metrics are S3 metrics - they are not compared to anything else. 
I guess that with time we'll polish them based on more understandings, but anyhow I expect them to be standalone.

**Comment by amnonh** [comment]
I mean that the dashboards are packed with metrics. It's ok to collect metrics and not show them, I guess we want to show that there is S3 activity but not sure that we need to show all those metrics

**Comment by regevran** [comment]
Please help me understand - not sure I know enough to smartly comment.
From my perspective:
The metrics we added are a "low-level" S3 metrics: How many bytes we read, how many we wrote, etc.
There are no "application" metrics at all. For example we don't count how many `SSTables` were backed up by the backup "application". We even don't count the number of times `S3-client` (that can serve number of "applications") failed.
We are on the very low-level of counters.

Are your concerns related to the fact that we have too-many counters that mostly will not be in use? because currently they are used only for backup and restore that happens once in a long period.

**Comment by amnonh** [comment]
The counters are there, the question, do we need them in the dashboard, my understanding, if we do, only a subset of them.

**Comment by regevran** [comment]
I see. 
What are the considerations when it comes to adding counters to the dashboard?

For me, having them all, at least for a while, will be beneficial as we don't know yet what we would use.

**Comment by amnonh** [comment]
@regevran how would you use any of them? you can always check specific information.
I'm still not sure, we want to make the dashboards easier for users, I think it's good to give an indication that
we upload data to S3, but not sure we need all those panels

**Comment by regevran** [comment]
Is it OK to add these 4:
```
scylla_s3_total_read_bytes
scylla_s3_total_write_bytes 
scylla_s3_total_read_latency_sec
scylla_s3_total_write_latency_sec
```
?

**Comment by amnonh** [comment]
I think it would be a good start

**Comment by mykaul** [comment]
> I see. What are the considerations when it comes to adding counters to the dashboard?
> 
> For me, having them all, at least for a while, will be beneficial as we don't know yet what we would use.

1. User valuable data - users would like to know what is (1) may be a cause for a performance/functionality issue and (2) what they might be paying for (which is slightly harder with the metrics here)
2. Developer/support data - we need to debug a performance/functionality issue.
3. Demo/show that 'it works'

So from that perspective, the above (https://github.com/scylladb/scylla-monitoring/issues/2103#issuecomment-2600783494 ) is only a reasonable fit for item no. 3 above.
I hope we can add other metrics (inc. no. of errors for example - do we have them somewhere?), histogram of request sizes , etc.

**Comment by robertbindar** [comment]
@amnonh I validated that https://github.com/scylladb/scylladb/issues/13904 is indeed fixed and we can get the s3 metrics on a ec2 instance with deb installed scylla. When your dashboards are up, we should be able to close this issue.
